test/evp_extra_test.c

```diff
@@ -1820,21 +1825,35 @@ static int test_pkey_ctx_fail_without_provider(int tst)
     if (!TEST_ptr(nullprov))
         goto err;
 
-    pctx = EVP_PKEY_CTX_new_from_name(tmpctx, tst == 0 ? "RSA" : "SM2", "");
-
-    /* RSA is not available via any provider so we expect this to fail */
-    if (tst == 0 && !TEST_ptr_null(pctx))
-        goto err;
-
     /*
-     * SM2 is always available because it is implemented via legacy codepaths
-     * and not in a provider at all. We expect this to pass.
-     * TODO(3.0): This can be removed once there are no more algorithms
-     * available via legacy codepaths
+     * We check for certain algos in the null provider.
+     * If an algo I expected to have a provider keymgmt, contructing an
+     * EVP_PKEY_CTX is expected to fail (return NULL).
+     * Otherwise, if it's expected to have legacy support, contructing an
+     * EVP_PKEY_CTX is expected to succeed (return non-NULL).
      */
-    if (tst == 1 && !TEST_ptr(pctx))
+    switch (tst) {
+    case 0:
+        keytype = "RSA";
+        expect_null = 1;
+        break;
+    case 1:
+        keytype = "SM2";
+        expect_null = 0; /* TODO: change to 1 when we have a SM2 keymgmt */
```

Comment:
Probably this part off the test can just be deleted when we have SM2 keymgmt. The point of the test is to check that algorithms with legacy codepaths work properly. Once SM2 keymgmt is in we won't have any algorithms in that category.

Reply:
Yes.  *Then!*  I assume you want this PR to pass the tests?  #12536 depends on this PR, not the other way around...

Reply:
I can ensure that's done in #12536

Reply:
Might point is the comment isn't really right. Rather than changing to 1 when we have SM2 keymgmt - we just delete it. But, its not  important. If we're going to delete it anyway we're talking about something that will only be in the codebase for a short period of time.

Reply:
What, so you only want to check RSA?  How arbitrary :laughing:

Reply:
Actually, thinking about it, I'd be tempted to just delete the whole test. I think it only exists because of the legacy case.

Reply:
Yes, I agree.